### PR TITLE
JPERF-531: Add support for wall-clock profiler

### PR DESCRIPTION
**Pull request by wyrzyk**
I'm using the implementation internally, but have no integration tests.

### CODE REVIEW

File: src/main/kotlin/com/atlassian/performance/tools/infrastructure/api/profiler/WallClockProfiler.kt [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    ): RemoteMonitoringProcess {
+        ssh.execute("./async-profiler/profiler.sh -e wall -i 9ms -f -b 160000000 start $pid")
+        return ProfilerProcess(pid)
+    }
+
+    private class ProfilerProcess(private val pid: Int) : RemoteMonitoringProcess {
+        private val flameGraphFile = "flamegraph.svg"
+
+        override fun stop(ssh: SshConnection) {
+            ssh.execute("./async-profiler/profiler.sh stop $pid -o svg > $flameGraphFile")
```
Comment (dagguh) [inline src/main/kotlin/com/atlassian/performance/tools/infrastructure/api/profiler/WallClockProfiler.kt]: Can we see the thread states in the flamegraph? If not, is there any way?

File: src/main/kotlin/com/atlassian/performance/tools/infrastructure/api/profiler/WallClockProfiler.kt [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    override fun start(
+        ssh: SshConnection,
+        pid: Int
+    ): RemoteMonitoringProcess {
+        ssh.execute("./async-profiler/profiler.sh -e wall -i 9ms -f -b 160000000 start $pid")
+        return ProfilerProcess(pid)
+    }
+
+    private class ProfilerProcess(private val pid: Int) : RemoteMonitoringProcess {
+        private val flameGraphFile = "flamegraph.svg"
```
Comment (dagguh) [inline src/main/kotlin/com/atlassian/performance/tools/infrastructure/api/profiler/WallClockProfiler.kt]: Is it legit to combine `AsyncProfiler` with `WallClockProfiler` at the same time?

File: CHANGELOG.md [hunk 3]
```diff
@@ -26,6 +26,9 @@ Dropping a requirement of a major version of a dependency is a new contract.
 ## [Unreleased]
 [Unreleased]: https://github.com/atlassian/infrastructure/compare/release-4.13.0...master
```
Comment (dagguh) [inline CHANGELOG.md]: FYI [4.14.0](https://packages.atlassian.com/maven-public/com/atlassian/performance/tools/infrastructure/4.14.0/) was [released](https://travis-ci.com/atlassian/infrastructure/builds/119456491) on 2019-07-17.

File: src/main/kotlin/com/atlassian/performance/tools/infrastructure/api/profiler/WallClockProfiler.kt (hunk 2 above)
Comment (wyrzyk) [reply src/main/kotlin/com/atlassian/performance/tools/infrastructure/api/profiler/WallClockProfiler.kt]: I'm not sure. I would not suggest doing that.

File: src/main/kotlin/com/atlassian/performance/tools/infrastructure/api/profiler/WallClockProfiler.kt (hunk 1 above)
Comment (wyrzyk) [reply src/main/kotlin/com/atlassian/performance/tools/infrastructure/api/profiler/WallClockProfiler.kt]: Nope, this way we gather all the threads. But you can use async-profiler for different `events` to track what you want (running threads/locks/context-switches).

File: CHANGELOG.md [hunk 4]
```diff
@@ -34,6 +37,7 @@ Dropping a requirement of a major version of a dependency is a new contract.
 
 
 [JPERF-523]: https://ecosystem.atlassian.net/browse/JPERF-523
+[JJPERF-531]: https://ecosystem.atlassian.net/browse/JPERF-531
```
Comment (mcmurphy2000) [inline CHANGELOG.md]: `[JJPERF-531]` -> `[JPERF-531]`

File: src/main/kotlin/com/atlassian/performance/tools/infrastructure/api/profiler/WallClockProfiler.kt [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+        ssh.execute("tar -xzf async-profiler-1.5-linux-x64.tar.gz -C async-profiler")
+        ssh.execute("sudo sh -c 'echo 1 > /proc/sys/kernel/perf_event_paranoid'")
+        ssh.execute("sudo sh -c 'echo 0 > /proc/sys/kernel/kptr_restrict'")
+    }
+
+    override fun start(
+        ssh: SshConnection,
+        pid: Int
+    ): RemoteMonitoringProcess {
+        ssh.execute("./async-profiler/profiler.sh -e wall -i 9ms -f -b 160000000 start $pid")
```
Comment (wyrzyk) [inline src/main/kotlin/com/atlassian/performance/tools/infrastructure/api/profiler/WallClockProfiler.kt]: I may need to adjust the buffer.

File: src/main/kotlin/com/atlassian/performance/tools/infrastructure/api/profiler/WallClockProfiler.kt (hunk 1 above)
Comment (wyrzyk) [reply src/main/kotlin/com/atlassian/performance/tools/infrastructure/api/profiler/WallClockProfiler.kt]: Can you explain the use case? I don't feel I have a full context to answer your question. 

Please check an example flame graph I've attached to the public ticket. This is the output you should expect. 

It'll aggregate state per multiple threads. You can use `async-profiler` with `-t` option to profile threads separately.

Comment (dagguh) [reply src/main/kotlin/com/atlassian/performance/tools/infrastructure/api/profiler/WallClockProfiler.kt]: The use case is bottleneck analysis.
Let's say 30% of actions are slow and bring Apdex down significantly. I can filter action metrics to narrow them down to the slow ones, then I can analyze their waterfalls and let's say most of them point to slow server. Then I need to profile server, but not it's entirety, but only the periods of time, when the slow actions were run. I don't want to include the stacktraces and thread states of the time where server was only serving quick actions.

Comment (wyrzyk) [reply src/main/kotlin/com/atlassian/performance/tools/infrastructure/api/profiler/WallClockProfiler.kt]: @dagguh do you mean something like heat maps?
https://medium.com/netflix-techblog/netflix-flamescope-a57ca19d47bb 
 http://www.brendangregg.com/HeatMaps/subsecondoffset.html

Comment (dagguh) [reply src/main/kotlin/com/atlassian/performance/tools/infrastructure/api/profiler/WallClockProfiler.kt]: Yes! This is brilliant!

**Comment by dagguh** [comment]
@wyrzyk should I wait for this or can I release `master` right now?

**Comment by dagguh** [comment]
I'm gonna go ahead and release

**Comment by dagguh** [comment]
GH features came in handy. The force-push diff showed me what the rebase did and the "files viewed" filter showed me that only the CHANGELOG has changed since last I saw.

**Pull request merged**